File: 1019-squares-of-a-sorted-array/squares-of-a-sorted-array.cpp
class Solution {
public:
    vector<int> sortedSquares(vector<int>& a) {
        int s = a.size();
        vector<int> neg;
        vector<int> pos;

        for(int i=0; i<s; i++){
            if (a[i]<0)
                neg.push_back(a[i]);
            else
                pos.push_back(a[i]);
        }

        if(neg.size()==0){
            for(int i=0; i<pos.size(); i++)
                pos[i]=pos[i]*pos[i];
            return pos;
        }

        if(pos.size()==0){
            for(int i=0; i<neg.size(); i++)
                neg[i]=neg[i]*neg[i];
            reverse(neg.begin(), neg.end());
            return neg;
        }

        int i=0, j=0, id=0;
        int n=neg.size();
        int m=pos.size();
        vector<int> res(n+m);

        // FIX: change loop variable name
        for(int i=0; i<n; i++)
            neg[i] = neg[i]*neg[i];

        reverse(neg.begin(),neg.end());

        for(int i=0; i<m; i++)
            pos[i] = pos[i] * pos[i];

        while(i<n && j<m){
            if (neg[i] <= pos[j]){
                res[id++] = neg[i++];
            }
            else{
                res[id++] = pos[j++];
            }
        }

        while(i<n){
            res[id++] = neg[i++];
        }

        while(j<m){
            res[id++] = pos[j++];
        }

        return res;
    }
};

File: 1019-squares-of-a-sorted-array/squares-of-a-sorted-array.py
class Solution(object):
    def sortedSquares(self, nums):
        size = len(nums)
        neg = []
        pos = []

        for num in nums:
            if num < 0:
                neg.append(num)
            else:
                pos.append(num)

        if len(neg) == 0:
            return [x * x for x in pos]

        if len(pos) == 0:
            return [x * x for x in neg][::-1]

        neg = [x * x for x in neg][::-1]
        pos = [x * x for x in pos]

        n, m = len(neg), len(pos)

        res = []

        i = 0
        j = 0

        while i < n and j < m:
            if neg[i] < pos[j]:
                res.append(neg[i])
                i += 1
            else:
                res.append(pos[j])
                j += 1

        while i < n:
            res.append(neg[i])
            i += 1

        while j < m:
            res.append(pos[j])
            j += 1

        return res